src/realm/object-store/sync/sync_session.cpp

```diff
@@ -1595,19 +1596,25 @@ SyncProgressNotifier::NotifierPackage::create_invocation(Progress const& current
     if (!is_download && snapshot_version > current_progress.snapshot_version)
         return [] {};
 
-    if (!is_streaming) {
-        // The initial download size we get from the server is the uncompacted
-        // size, and so the download may complete before we actually receive
-        // that much data. When that happens, transferrable will drop and we
-        // need to use the new value instead of the captured one.
+    // only invoke the callback on registration if is in active data transfer
+    if (is_download && initial_registration && progress_estimate >= 1)
```

Comment:
why do we need to check the progress estimate? don't we want to invoke the callback if in steady state already?

Reply:
ah, I think I get it now. I was thinking about query version 0, but that hits a different code path. This is needed if progress reaches 1.0 and then the session is paused and resumed. we don't want to immediately report 1.0 since there may be history scanning (and more data incoming). Don't we want to do this only for flx though? And if so, one other potential way to fix it is to reset `m_current_progress` when the session is paused/started.

Reply:
@kiburtse any updates on this?

Reply:
hypothetically, this applies also to the running session after the batch was synced, next callback registration will fail the same way. I'm double-checking if this is actually the usecase. Maybe we should support it also. 

Clearing `m_current_progress` is problematic, since it is used not in the NotifierPackage (callback wrapper), but in SyncProgressNotifier. So it'd have an effect of every other callback (upload non-streaming included). Also, it would couple the session state with the notifier, so far notifier only behaves based on values reported to the user. I don't think it's the right approach. 

Also, and i don't see how it's flx specific. Essentially, pbs behaves the same way - at least from the observed behavior on the user side. I think it's worth it to keep them in sync implementation wise. 

Anyway, let me check if we should check not just `initial_registration` but also essentially every update for initial `create_invocation` for the callback, since update from upload after full download would break it probably.

Reply:
> hypothetically, this applies also to the running session after the batch was synced, next callback registration will fail the same way. I'm double-checking if this is actually the usecase. Maybe we should support it also.

Indeed, but it will report 1.0 when the next download message is received otherwise (unless a new bootstrap starts)

> Clearing m_current_progress is problematic, since it is used not in the NotifierPackage (callback wrapper), but in SyncProgressNotifier. So it'd have an effect of every other callback (upload non-streaming included). Also, it would couple the session state with the notifier, so far notifier only behaves based on values reported to the user. I don't think it's the right approach.

Agree, it was mostly to see if you considered it.

> Also, and i don't see how it's flx specific. Essentially, pbs behaves the same way - at least from the observed behavior on the user side. I think it's worth it to keep them in sync implementation wise.

For pbs we don't get an estimate (the client computes it), so I think at registration, if there was at least one a download message received previously, then the callback would fire right? This changes now.

Reply:
> > hypothetically, this applies also to the running session after the batch was synced, next callback registration will fail the same way. I'm double-checking if this is actually the usecase. Maybe we should support it also.
> 
> Indeed, but it will report 1.0 when the next download message is received otherwise (unless a new bootstrap starts)
> 
ok, i haven't figured out what could the test scenario be for this, but just in case i've added additional test for SyncProgressNotifier to check for this, also including the situation where there are bunch of uploads in between. I had to tweak the logic in ProgressNotifier a bit for this.

> > Also, and i don't see how it's flx specific. Essentially, pbs behaves the same way - at least from the observed behavior on the user side. I think it's worth it to keep them in sync implementation wise.
> 
> For pbs we don't get an estimate (the client computes it), so I think at registration, if there was at least one a download message received previously, then the callback would fire right? This changes now.
>
yes, you are right - that is what changes. I've put sample testcase in the pr description. Initial synchronization or reconnect and initial sync don't change for pbs, and for flx it just works now. The case where "one a download message received previously, then the callback would fire" isn't useful, especially since we now have progress estimate, which is always 1.0 for this case. That was the complaint from sdks in the nutshell, according to my understanding. Now we allow essentially to get the progress if it would happen after initial sync. I think it's a fix, but in general whole non-streaming download callback is not very useful, since every download message in steady state is its own batch (with progress 1.0 for flx, and 0 downloadable_bytes for pbs). I don't believe we can fix it any further without some more fundamental changes.